File: pixel_refine_desktop/enhance_stack/core/algorithm/alignment/alignment_tile/refinement.cpp
#include "refinement.hpp"
#include "cost_function.hpp"

#include <cmath>
#include <opencv2/imgproc.hpp>
#include <opencv2/tracking.hpp>


// ============================================================================
// OPTIMIZATION 1: Fast bicubic interpolation tanpa overhead cv::remap
// ============================================================================

// Bicubic kernel weights (pre-computed untuk performance)
static inline float cubic_weight(float x) {
  // Catmull-Rom spline (standard bicubic)
  x = std::fabs(x);
  if (x <= 1.0f) {
    return 1.5f * x * x * x - 2.5f * x * x + 1.0f;
  } else if (x < 2.0f) {
    return -0.5f * x * x * x + 2.5f * x * x - 4.0f * x + 2.0f;
  }
  return 0.0f;
}

// OPTIMIZED: Direct bicubic interpolation (10-20x lebih cepat dari cv::remap)
static inline float bicubic_at_optimized(const cv::Mat &img, float x, float y) {
  // Boundary check
  if (x < 1.0f || y < 1.0f || x >= img.cols - 2.0f || y >= img.rows - 2.0f) {
    // Fallback to bilinear untuk edge cases
    int ix = static_cast<int>(x);
    int iy = static_cast<int>(y);

    if (ix < 0 || iy < 0 || ix >= img.cols - 1 || iy >= img.rows - 1) {
      return 0.0f;
    }

    float fx = x - ix;
    float fy = y - iy;

    const float *row0 = img.ptr<float>(iy);
    const float *row1 = img.ptr<float>(iy + 1);

    float top = row0[ix] * (1.0f - fx) + row0[ix + 1] * fx;
    float bottom = row1[ix] * (1.0f - fx) + row1[ix + 1] * fx;

    return top * (1.0f - fy) + bottom * fy;
  }

  // Integer part dan fractional part
  int ix = static_cast<int>(std::floor(x));
  int iy = static_cast<int>(std::floor(y));
  float fx = x - ix;
  float fy = y - iy;

  // Pre-compute weights (avoiding repeated calculations)
  float wx[4], wy[4];
  wx[0] = cubic_weight(fx + 1.0f);
  wx[1] = cubic_weight(fx);
  wx[2] = cubic_weight(1.0f - fx);
  wx[3] = cubic_weight(2.0f - fx);

  wy[0] = cubic_weight(fy + 1.0f);
  wy[1] = cubic_weight(fy);
  wy[2] = cubic_weight(1.0f - fy);
  wy[3] = cubic_weight(2.0f - fy);

  // Bicubic interpolation (4x4 neighborhood)
  float sum = 0.0f;

  // Manual unroll untuk better performance
  const float *rows[4];
  rows[0] = img.ptr<float>(iy - 1);
  rows[1] = img.ptr<float>(iy);
  rows[2] = img.ptr<float>(iy + 1);
  rows[3] = img.ptr<float>(iy + 2);

  // Compute weighted sum
  for (int j = 0; j < 4; ++j) {
    float row_sum = 0.0f;
    const int base_x = ix - 1;

    // Unrolled inner loop
    row_sum += rows[j][base_x] * wx[0];
    row_sum += rows[j][base_x + 1] * wx[1];
    row_sum += rows[j][base_x + 2] * wx[2];
    row_sum += rows[j][base_x + 3] * wx[3];

    sum += row_sum * wy[j];
  }

  return sum;
}

// ============================================================================
// OPTIMIZATION 2: Vectorized SAD calculation untuk final validation
// ============================================================================

static inline float compute_sad_with_bicubic_avx(const cv::Mat &ref_layer,
                                                 const cv::Mat &comp_layer,
                                                 int x, int y, float flow_x,
                                                 float flow_y, int tile_w,
                                                 int tile_h) {
  float sad_total = 0.0f;

  // Process rows dengan bicubic interpolation
  for (int r = 0; r < tile_h; ++r) {
    const float *p_ref = ref_layer.ptr<float>(y + r, x);
    const float comp_y = y + flow_y + r;

    // Vectorize the inner loop dengan manual accumulation
    float row_sad = 0.0f;

#pragma omp simd reduction(+ : row_sad)
    for (int c = 0; c < tile_w; ++c) {
      float ref_val = p_ref[c];
      float comp_val = bicubic_at_optimized(comp_layer, x + flow_x + c, comp_y);
      row_sad += std::fabs(ref_val - comp_val);
    }

    sad_total += row_sad;
  }

  return sad_total / (tile_w * tile_h);
}

// ============================================================================
// OPTIMIZATION 3: Parabolic Fitting untuk Fast Sub-pixel Refinement (HYBRID)
// ============================================================================

/**
 * Parabolic Fitting pada 3x3 grid correlation surface
 * Evaluates 9 points dan fit parabola untuk mendapatkan sub-pixel peak
 * Return: {refined_flow, confidence_score}
 */
static inline std::pair<cv::Point2f, float>
parabolic_refinement(const cv::Mat &ref_layer, const cv::Mat &comp_layer, int x,
                     int y, int dx, int dy, int tile_w, int tile_h) {
  constexpr float TILE_AREA_INV =
      1.0f / (256.0f); // Pre-compute untuk tile 16x16

  // Setup reference tile rows (cached)
  std::vector<const float *> ref_rows(tile_h);
  for (int r = 0; r < tile_h; ++r) {
    ref_rows[r] = ref_layer.ptr<float>(y + r, x);
  }

  // Evaluate 9 points pada 3x3 grid
  float costs[9]; // Row-major: [-1,-1], [0,-1], [1,-1], [-1,0], [0,0], [1,0],
                  // [-1,1], [0,1], [1,1]
  int eval_idx = 0;

  for (int ddy = -1; ddy <= 1; ++ddy) {
    for (int ddx = -1; ddx <= 1; ++ddx) {
      const int test_dx = dx + ddx;
      const int test_dy = dy + ddy;

      // Boundary check
      if (x + test_dx < 0 || y + test_dy < 0 ||
          x + test_dx + tile_w > comp_layer.cols ||
          y + test_dy + tile_h > comp_layer.rows) {
        costs[eval_idx] = std::numeric_limits<float>::max();
      } else {
        // Calculate cost untuk posisi ini
        float total_cost = 0.0f;
        for (int r = 0; r < tile_h; ++r) {
          const float *p_ref = ref_rows[r];
          const float *p_comp =
              comp_layer.ptr<float>(y + test_dy + r, x + test_dx);

          for (int c = 0; c < tile_w; ++c) {
            total_cost += std::fabs(p_ref[c] - p_comp[c]);
          }
        }
        costs[eval_idx] = total_cost * TILE_AREA_INV;
      }
      eval_idx++;
    }
  }

  // Parabolic fitting pada costs[4] (center) dengan 8 neighbors
  // Gunakan least-squares fit untuk 2D parabola: f(u,v) = a + b*u + c*v + d*u^2
  // + e*v^2 + f*u*v Simplified: Fit 1D parabola untuk each axis

  float center_cost = costs[4]; // [0, 0]
  float left_cost = costs[3];   // [-1, 0]
  float right_cost = costs[5];  // [1, 0]
  float top_cost = costs[1];    // [0, -1]
  float bottom_cost = costs[7]; // [0, 1]

  // Parabolic fit: f(x) = a + b*x + c*x^2
  // Min occurs at x = -b / (2*c)

  // X-direction fit (menggunakan left, center, right)
  float delta_x = 0.0f;
  {
    float c_coeff = (right_cost + left_cost - 2.0f * center_cost) / 2.0f;
    if (std::fabs(c_coeff) > 1e-6f) {
      float b_coeff = (right_cost - left_cost) / 2.0f;
      delta_x = -b_coeff / (2.0f * c_coeff);
      // Clamp untuk mencegah outlier
      delta_x = std::max(-0.5f, std::min(0.5f, delta_x));
    }
  }

  // Y-direction fit (menggunakan top, center, bottom)
  float delta_y = 0.0f;
  {
    float c_coeff = (bottom_cost + top_cost - 2.0f * center_cost) / 2.0f;
    if (std::fabs(c_coeff) > 1e-6f) {
      float b_coeff = (bottom_cost - top_cost) / 2.0f;
      delta_y = -b_coeff / (2.0f * c_coeff);
      // Clamp untuk mencegah outlier
      delta_y = std::max(-0.5f, std::min(0.5f, delta_y));
    }
  }

  // Compute confidence sebagai inverse dari curvature (sharpness of peak)
  float curvature_x = (right_cost + left_cost - 2.0f * center_cost) / 2.0f;
  float curvature_y = (bottom_cost + top_cost - 2.0f * center_cost) / 2.0f;
  float curvature = std::max(std::fabs(curvature_x), std::fabs(curvature_y));

  // Confidence: semakin sharp peak (curvature besar), semakin confident
  // Normalisasi: curvature range [0.001, 0.1] -> confidence [0.1, 0.9]
  float confidence = 0.5f;
  if (curvature > 1e-6f) {
    confidence = std::min(0.9f, 0.1f + (std::log10(curvature) + 3.0f) * 0.2f);
    confidence = std::max(0.1f, confidence);
  }

  cv::Point2f refined_flow(static_cast<float>(dx) + delta_x,
                           static_cast<float>(dy) + delta_y);

  return {refined_flow, confidence};
}

// ============================================================================
// OPTIMIZATION 4: Main optimized subpixel refinement (HYBRID APPROACH)
// ============================================================================

constexpr int MAX_STACK_TILE_SIZE = 64;

cv::Point2f subpixel_refinement(const cv::Mat &ref_layer,
                                const cv::Mat &comp_layer, int x, int y, int dx,
                                int dy, int tile_w, int tile_h) {
  // Parameters
  constexpr double MIN_CONFIDENCE_RESPONSE = 0.30;
  constexpr double MAX_CONFIDENCE_RESPONSE = 0.95;
  constexpr float MAX_FINAL_SAD_PER_PIXEL = 0.05f;

  // Boundaries
  const int ref_max_x = ref_layer.cols - tile_w;
  const int ref_max_y = ref_layer.rows - tile_h;
  const int comp_max_x = comp_layer.cols - tile_w;
  const int comp_max_y = comp_layer.rows - tile_h;

  // Early boundary check
  if (x < 0 || y < 0 || x > ref_max_x || y > ref_max_y) {
    return cv::Point2f(static_cast<float>(dx), static_cast<float>(dy));
  }

  // =========================================================================
  // STEP 1: Integer search 3x3 (OPTIMIZED)
  // =========================================================================
  float min_cost = std::numeric_limits<float>::max();
  int best_integer_dx = dx;
  int best_integer_dy = dy;

  // OPTIMIZATION 5: Stack Array replacement for std::vector
  const float *ref_rows_stack[MAX_STACK_TILE_SIZE];

  // Fallback jika tile terlalu besar (sangat jarang terjadi untuk refinement)
  std::vector<const float *> ref_rows_heap;
  const float **ref_rows_ptr;
  if (tile_h <= MAX_STACK_TILE_SIZE) {
    for (int r = 0; r < tile_h; ++r) {
      ref_rows_stack[r] = ref_layer.ptr<float>(y + r, x);
    }
    ref_rows_ptr = ref_rows_stack;
  } else {
    // Fallback ke Heap jika ukuran tile tidak wajar
    ref_rows_heap.resize(tile_h);
    for (int r = 0; r < tile_h; ++r) {
      ref_rows_heap[r] = ref_layer.ptr<float>(y + r, x);
    }
    ref_rows_ptr = ref_rows_heap.data();
  }

  // Pre-calculate stride untuk akses cepat comp_layer
  const size_t comp_step = comp_layer.step1();

  // 3x3 search (ddy: -1 to 1, ddx: -1 to 1)
  for (int ddy = -1; ddy <= 1; ++ddy) {
    const int current_dy = dy + ddy;
    const int comp_y = y + current_dy;

    // Row boundary check
    if (comp_y < 0 || comp_y > comp_max_y)
      continue;

    // Ambil pointer baris pertama comp untuk current_dy
    // Kita bisa menggeser pointer ini ke bawah menggunakan comp_step
    const float *comp_row_ptr_base = comp_layer.ptr<float>(comp_y);

    for (int ddx = -1; ddx <= 1; ++ddx) {
      const int current_dx = dx + ddx;
      const int comp_x = x + current_dx;

      // Column boundary check
      if (comp_x < 0 || comp_x > comp_max_x)
        continue;

      // OPTIMIZATION 6: Row-wise ZSAD dengan Pointer Arithmetic
      float total_cost = 0.0f;

      // Manual loop tanpa memanggil .ptr() berulang kali
      for (int r = 0; r < tile_h; ++r) {
        // Pointer reference sudah di-cache
        const float *p_ref = ref_rows_ptr[r];

        // Pointer comp dihitung manual: base + (row * step) + offset_x
        const float *p_comp = comp_row_ptr_base + (r * comp_step) + comp_x;

        // Fungsi kalkulasi cost eksternal (pastikan fungsi ini di-inline oleh
        // compiler)
        total_cost += calculate_fine_analysis(p_ref, p_comp, tile_w);
      }

      if (total_cost < min_cost) {
        min_cost = total_cost;
        best_integer_dx = current_dx;
        best_integer_dy = current_dy;
      }
    }
  }

  const cv::Point2f integer_flow(static_cast<float>(best_integer_dx),
                                 static_cast<float>(best_integer_dy));

  // =========================================================================
  // STEP 2: Parabolic Fitting ONLY (untuk kecepatan maksimum)
  // =========================================================================
  auto [parabolic_flow, parabolic_confidence] =
      parabolic_refinement(ref_layer, comp_layer, x, y, best_integer_dx,
                           best_integer_dy, tile_w, tile_h);

  // Selalu return parabolic result untuk kecepatan (skip ECC sepenuhnya)
  // Parabolic sudah cukup akurat untuk mayoritas cases dan 10x lebih cepat dari
  // ECC
  return parabolic_flow;
}

File: pixel_refine_desktop/enhance_stack/core/algorithm/alignment/alignment_tile/refinement.py
"""
Refinement - Taichi GPU Implementation
======================================
GPU-accelerated subpixel refinement functions for optical flow.
"""

import numpy as np

try:
    import taichi as ti
    import taichi.math as tm

    TAICHI_AVAILABLE = True
except ImportError:
    TAICHI_AVAILABLE = False
    ti = None
    tm = None

try:
    from . import cost_function
except ImportError:
    # Facilitate standalone tests
    try:
        import cost_function
    except ImportError:
        cost_function = None

try:
    from ...taichi_algorithm import common
except ImportError:
    # Facilitate standalone tests
    try:
        import common
    except ImportError:
        common = None

# ============================================================================
# Constants
# ============================================================================
MAX_STACK_TILE_SIZE = 64


# ============================================================================
# Taichi Kernels
# ============================================================================

if TAICHI_AVAILABLE:

    @ti.kernel
    def _subpixel_refinement_parabolic_kernel(
        ref_layer: ti.types.ndarray(),
        comp_layer: ti.types.ndarray(),
        result_out: ti.types.ndarray(),
        tile_x: int,
        tile_y: int,
        dx: int,
        dy: int,
        tile_w: int,
        tile_h: int,
        h: int,
        w: int,
    ):
        """
        Integrated Subpixel Refinement:
        1. 3x3 Integer search using ZMCL cost.
        2. 3x3 Parabolic fitting using SAD cost around the best integer position.
        Matching C++ logic flow.
        """
        # Step 1: Integer search (3x3 ZMCL)
        min_cost = 1e10
        best_dx = dx
        best_dy = dy

        for ddy in ti.static(range(-1, 2)):
            for ddx in ti.static(range(-1, 2)):
                test_dx = dx + ddx
                test_dy = dy + ddy

                if (
                    tile_x + test_dx >= 0
                    and tile_y + test_dy >= 0
                    and tile_x + test_dx + tile_w <= w
                    and tile_y + test_dy + tile_h <= h
                ):
                    cost = cost_function.compute_zmsad_cost(
                        ref_layer,
                        comp_layer,
                        tile_y,
                        tile_x,
                        tile_y + test_dy,
                        tile_x + test_dx,
                        tile_h,
                        tile_w,
                    )
                    if cost < min_cost:
                        min_cost = cost
                        best_dx = test_dx
                        best_dy = test_dy

        # Step 2: Parabolic Refinement using SAD (3x3 grid around best_dx, best_dy)
        costs = ti.Vector([0.0] * 9)
        eval_idx = 0
        for ddy in ti.static(range(-1, 2)):
            for ddx in ti.static(range(-1, 2)):
                test_dx = best_dx + ddx
                test_dy = best_dy + ddy

                if (
                    tile_x + test_dx < 0
                    or tile_y + test_dy < 0
                    or tile_x + test_dx + tile_w > w
                    or tile_y + test_dy + tile_h > h
                ):
                    costs[eval_idx] = 1e10
                else:
                    costs[eval_idx] = cost_function.compute_zmssd_cost(
                        ref_layer,
                        comp_layer,
                        tile_y,
                        tile_x,
                        tile_y + test_dy,
                        tile_x + test_dx,
                        tile_h,
                        tile_w,
                    )
                eval_idx += 1

        # Parabolic fitting on the SAD surface
        center_cost = costs[4]
        left_cost = costs[3]
        right_cost = costs[5]
        top_cost = costs[1]
        bottom_cost = costs[7]

        delta_x = 0.0
        c_coeff_x = (right_cost + left_cost - 2.0 * center_cost) / 2.0
        if ti.abs(c_coeff_x) > 1e-6:
            b_coeff_x = (right_cost - left_cost) / 2.0
            delta_x = tm.clamp(-b_coeff_x / (2.0 * c_coeff_x), -0.5, 0.5)

        delta_y = 0.0
        c_coeff_y = (bottom_cost + top_cost - 2.0 * center_cost) / 2.0
        if ti.abs(c_coeff_y) > 1e-6:
            b_coeff_y = (bottom_cost - top_cost) / 2.0
            delta_y = tm.clamp(-b_coeff_y / (2.0 * c_coeff_y), -0.5, 0.5)

        # Confidence based on curvature (matching C++)
        curvature = ti.max(ti.abs(c_coeff_x), ti.abs(c_coeff_y))
        confidence = 0.5
        if curvature > 1e-6:
            confidence = tm.clamp(
                0.1 + (ti.log(curvature) / ti.log(10.0) + 3.0) * 0.2, 0.1, 0.9
            )

        result_out[0] = float(best_dx) + delta_x
        result_out[1] = float(best_dy) + delta_y
        result_out[2] = confidence

    @ti.kernel
    def _bicubic_subpixel_refinement_kernel(
        ref_layer: ti.types.ndarray(),
        comp_layer: ti.types.ndarray(),
        result_out: ti.types.ndarray(),
        tile_x: int,
        tile_y: int,
        dx: float,
        dy: float,
        tile_w: int,
        tile_h: int,
        h: int,
        w: int,
    ):
        """
        Subpixel refinement using iterative bicubic sampling.
        Performs a local grid search for higher precision.
        """
        best_vx = dx
        best_vy = dy
        best_cost = 1e10

        # Initial step for local refinement (e.g. 0.5 pixel)
        step = 0.5

        # 3 iterations of local grid search (±step)
        for _ in ti.static(range(4)):
            best_local_vx = best_vx
            best_local_vy = best_vy

            for ddy in ti.static(range(-1, 2)):
                for ddx in ti.static(range(-1, 2)):
                    vx = best_vx + float(ddx) * step
                    vy = best_vy + float(ddy) * step

                    # Boundary check for tile center
                    if not (
                        tile_x + vx < 1.0
                        or tile_y + vy < 1.0
                        or tile_x + vx + tile_w > w - 2
                        or tile_y + vy + tile_h > h - 2
                    ):
                        # Evaluate SAD cost at (vx, vy) using bicubic interpolation
                        total_cost = 0.0
                        for r, c in ti.ndrange(tile_h, tile_w):
                            ref_val = ref_layer[tile_y + r, tile_x + c]
                            comp_val = common.bicubic_at(
                                comp_layer,
                                float(tile_x + c) + vx,
                                float(tile_y + r) + vy,
                            )
                            total_cost += ti.abs(ref_val - comp_val)

                        cost = total_cost / float(tile_w * tile_h)
                        if cost < best_cost:
                            best_cost = cost
                            best_local_vx = vx
                            best_local_vy = vy

            best_vx = best_local_vx
            best_vy = best_local_vy
            step *= 0.5  # Reduce step size for next iteration

        result_out[0] = best_vx
        result_out[1] = best_vy
        result_out[2] = 0.9  # High confidence for bicubic


# ============================================================================
# Python API (matching C++ function signatures)
# ============================================================================


def parabolic_refinement(
    ref_layer: np.ndarray,
    comp_layer: np.ndarray,
    x: int,
    y: int,
    dx: int,
    dy: int,
    tile_w: int,
    tile_h: int,
) -> tuple:
    """
    Parabolic fitting for subpixel refinement.
    Matching C++ parabolic_refinement().

    Args:
        ref_layer: Reference image (2D float32)
        comp_layer: Comparison image (2D float32)
        x, y: Tile top-left position in reference
        dx, dy: Integer displacement to refine
        tile_w, tile_h: Tile dimensions

    Returns:
        Tuple of (refined_flow, confidence)
        refined_flow: (x, y) subpixel displacement
        confidence: 0.0-1.0 confidence score
    """
    if not TAICHI_AVAILABLE:
        raise ImportError("Taichi not available")

    h, w = ref_layer.shape[:2]

    ref_gpu = ti.ndarray(ti.f32, shape=(h, w))
    comp_gpu = ti.ndarray(ti.f32, shape=(h, w))
    result_out = ti.ndarray(ti.f32, shape=(3,))

    ref_gpu.from_numpy(np.ascontiguousarray(ref_layer, dtype=np.float32))
    comp_gpu.from_numpy(np.ascontiguousarray(comp_layer, dtype=np.float32))

    return (result[0], result[1]), result[2]


def subpixel_refinement(
    ref_layer: np.ndarray,
    comp_layer: np.ndarray,
    x: int,
    y: int,
    dx: int,
    dy: int,
    tile_w: int,
    tile_h: int,
) -> tuple:
    """
    Combined integer search + parabolic refinement.
    Matching C++ subpixel_refinement().

    Args:
        ref_layer: Reference image (2D float32)
        comp_layer: Comparison image (2D float32)
        x, y: Tile top-left position in reference
        dx, dy: Initial integer displacement
        tile_w, tile_h: Tile dimensions

    Returns:
        Tuple (refined_dx, refined_dy) with subpixel accuracy
    """
    if not TAICHI_AVAILABLE:
        raise ImportError("Taichi not available")

    h, w = ref_layer.shape[:2]

    # Initialize GPU buffers if they don't exist
    ref_gpu = ti.ndarray(ti.f32, shape=(h, w))
    comp_gpu = ti.ndarray(ti.f32, shape=(h, w))
    result_out = ti.ndarray(ti.f32, shape=(3,))

    ref_gpu.from_numpy(np.ascontiguousarray(ref_layer, dtype=np.float32))
    comp_gpu.from_numpy(np.ascontiguousarray(comp_layer, dtype=np.float32))

    # Boundary check
    if x < 0 or y < 0 or x + tile_w > w or y + tile_h > h:
        return (float(dx), float(dy))

    # Combined Integer Search + Parabolic Refinement in one kernel
    _subpixel_refinement_parabolic_kernel(
        ref_gpu, comp_gpu, result_out, x, y, dx, dy, tile_w, tile_h, h, w
    )

    result = result_out.to_numpy()
    return (result[0], result[1])


def subpixel_refinement_gpu(
    ref_gpu,
    comp_gpu,
    search_result_gpu,
    refine_result_gpu,
    x: int,
    y: int,
    dx: int,
    dy: int,
    tile_w: int,
    tile_h: int,
    h: int,
    w: int,
) -> None:
    """
    GPU-native version: Performs refinement directly on GPU buffers.
    Results written to search_result_gpu and refine_result_gpu.
    """
    if not TAICHI_AVAILABLE:
        raise ImportError("Taichi not available")

    # Integrated subpixel refinement
    _subpixel_refinement_parabolic_kernel(
        ref_gpu,
        comp_gpu,
        refine_result_gpu,
        x,
        y,
        dx,
        dy,
        tile_w,
        tile_h,
        h,
        w,
    )
